Documentation for TLS certificate rotation feature (#2165)
neo4j/docs-operations: modules/ROOT/pages/security/ssl-framework.adoc

To avoid that scenario, make sure this setting is only enabled for a short term duration.
====

[role=label--new-2025.03]
[[certificate-rotation]]

> == Certificates rotation

Author: I don't think so. We are already on the SSL Framework page

It is considered best practice to use certificates with reasonably short duration.
This, however, requires the periodic rotation of certificates whereby old certificates are removed and the new ones are installed.
Previous versions of Neo4j required a database restart for changes to be applied.
Starting from 2025.03, new certificates can be rotated in, and SSL configuration can be updated without requiring a restart.
This reduces undesirable effects of transient loss of cluster members.

> Following are the steps of newly introduced certificates rotation.

Reviewer: ```suggestion
The following steps outline the process for certificates rotation.
```

. Enable the dynamic reloading of certificates on all cluster members.
It is best to do this when the cluster is deployed as changing this configuration requires a restart:

[source, properties]
----

> dbms.security.tls_reload=true (default is false)

Reviewer: One more question: Do we need to mention this setting above https://neo4j.com/docs/operations-manual/current/security/ssl-framework/#ssl-configuration? where we talk about SSL configuration settings

> dbms.security.tls_reload=true (default is false)

Author: We could mention it on https://neo4j.com/docs/operations-manual/current/clustering/setup/deploy/ but I don't think we should - it's an optional feature and we don't mention SSL setup anywhere else on that page. I guess same applies to Docker and Kubernetes sections.

I don't think it needs to go above the other settings - again this is an optional feature and isn't required knowledge to fully understand and deploy the SSL configuration

----

> . Replace old certificates either by overwriting on the filesystem or by copying them to a new location and updating the required SSL configuration for each effected scope.

Reviewer: ```suggestion
. Replace old certificates either by overwriting them on the filesystem or by copying them to a new location.
Then update the required SSL configuration for each effected scope.

```

> Both certificates may exist on the filesystem but only one can be referenced in the configuration.

Reviewer: ```suggestion
New and old certificates may exist on the filesystem but only one can be referenced in the configuration.
```

> New and old certificates may exist on the filesystem but only one can be referenced in the configuration.

Reviewer: ```suggestion
New and old certificates may co-exist on the filesystem, but only one can be referenced in the configuration.
```

New certificates need to be copied to all cluster members as required.

. Make necessary changes to any of the SSL configuration and/or replace certificates for effected scopes. 

. Connect to each cluster member in turn with Cypher Shell using a <<ssl-bolt-connect,`bolt` URI scheme>> and run the reload procedure:

[source]
----

> dbms.reloadTLSCertificates()

Reviewer: Is this a new procedure? Our test didn't show yet that we miss a procedure on our page https://neo4j.com/docs/operations-manual/current/procedures/. But looks like we need to add in this procedure there

Author: Yes it's a new procedure but not merged yet. I'll add it to this page too once we agree the name.

Author: added

----

> . New settings will take effect immediately, however existing connections will not be pre-emptively terminated.

Reviewer: I think I need more details here. Does this mean that new connections will use new certificates, but the existing connections (established before the update) will continue using the old certificates until they expire?

Author: Yes exactly. It will not terminate existing connection. But any new connections will use the new certificates

Author: Even expired certificates won't cause a connection to terminate. The certificates are only used during the connection negotiation stage

> . New settings will take effect immediately, however existing connections will not be pre-emptively terminated.

Reviewer: ```suggestion
. New settings will take effect immediately; however, existing connections will not be preemptively terminated.
This means that while new connections will use new certificates, the existing connections (established before the update) will continue using the old certificates.
Even if a certificate expires, active connections remain unaffected because the certificates are only used during the initial connection handshake.
```
